initial implementation of goto types and separating annotations

Acepie opened the pull request:
<!-- Don't forget to update the CHANGELOG! -->
An initial implementation for making annotations distinct from the thing being annotated. Currently in most places where the user hovers over an annotation it either doesn't work or hovers the larger object being annotated (function head, constant, custom type). This change is the first part of separating that out and allowing for operations on the located annotation. Currently that means hovering will display the type itself, completions will always be type completions (this seems to be exactly the same as current in all cases), and goto definition now will go to where the type is declared. There are a couple known gaps here but overall its still an improvement so want to start with this and get feedback

Known things to improve:

- Type aliases seem to be stored in the ast directly as the type being aliased meaning that hover/goto will use the underlying type
- ~Hovering over a type won't show documentation~
- ~Because of how locations/labels for CustomType constructor args are implemented hovering over the arg label will be treated as hovering over the annotation~
  -  ~This is probably fixable but honestly I kinda like it so wanna get thoughts~

Acepie commented on compiler-core/src/language_server/compiler.rs:
```diff
@@ -124,6 +130,12 @@ where
             // Create the source information
             let path = module.src_path.to_string();
+            // strip canonicalised windows prefix
+            #[cfg(target_family = "windows")]
```
ran into this while testing on my other computer. Only causes issues with local deps (I'm guessing due to calling canonicalise similar to https://github.com/gleam-lang/gleam/pull/2842). Could pull this out into a separate thing if preferred

Acepie replied:
Also, not sure if you want me to pull this out if this pr won't get into the 1.1 release (I'm fine with whatever but lmk what you think. I think the fix here really should be in 1.1 but the rest makes sense to leave out)

lpil commented on compiler-core/src/type_.rs:
```diff
@@ -134,6 +134,28 @@ impl Type {
+        }
+    }
+
+    pub fn nested_type(&self) -> Option<Arc<Self>> {
```
What's a nested type?

Could you document these three functions please 🙏

Acepie replied:
I'm realizing inner_type is probably a better name for this. I just needed a way to follow the Var to the underlying type so we had proper linking. Lemme rename and also add comments

lpil commented on compiler-core/src/language_server/tests/definition.rs:
```diff
@@ -632,3 +632,129 @@ fn main() {
+            }
+        })
+    )
+}
```
Worth having a test for path deps?

Acepie replied:
lemme add one since it'll be easy though i think? it'll use all the same code paths

lpil commented:
> Because of how locations/labels for CustomType constructor args are implemented hovering over the arg label will be treated as hovering over the annotation 

Does that mean if I have this code (`][ is cursor`)

```gleam
pub type Thing {
  Thing(value: ][ String)
}
```
then it shows documentation for `type String`?

Acepie commented:
Yeah because of how the parser sets the location the entire arg is String. We could add some additional information to the AST if we wanna split it but I think the way it seemed to me rn "value" is treated as a label in the AST

Acepie commented:
Huh actually, I wanted to test this out after I typed that and I guess somewhere along the line that statement became no longer true and I'm not sure why. Wondering if maybe I had a bug in an earlier version or if there is something I am not remembering rn

Acepie commented:
Ok yeah the actual behavior is that the label has no hover (which matches the previous behavior). I think i had a bug at some middle phase that meant i was checking the wrong location and fixed it when I switched to do the TypeAst method of finding the location

lpil reviewed (commented):
Smashing it as always! Thank you!

Left some v small questions

Acepie commented:
Ok figured out how to get the documentation on hover for custom types. Ultimately ended up adding documentation to TypeConstructors and then passing it through and looking up the type on hover

<img width="1047" alt="Screenshot 2024-04-09 at 8 24 06 AM" src="https://github.com/gleam-lang/gleam/assets/5996838/733dac83-8053-465c-bb36-c92d33104338">

Acepie commented:
I pulled the windows path fix into https://github.com/gleam-lang/gleam/pull/2941

lpil commented on compiler-core/src/build.rs:
```diff
@@ -296,10 +296,14 @@ pub enum Located<'a> {
     ModuleStatement(&'a TypedDefinition),
     FunctionBody(&'a TypedFunction),
     Arg(&'a TypedArg),
+    Annotation(&'a TypeAst, std::sync::Arc<type_::Type>),
```
What do you think of just having the location here? It seems that's all we use and I can't think of anything else that we could take from this data structure that would be useful.

Acepie replied:
Yeah I think replacing the TypeAst with just the location is reasonable. The type itself we need for figuring out the type constructor

lpil replied:
Sorry yes that's what I meant, in place of the AST

lpil commented on compiler-core/src/language_server/engine.rs:
```diff
@@ -154,7 +154,12 @@ where
             };
 
-            let location = match node.definition_location() {
+            let location = match node.definition_location(node.annotation_type().and_then(|t| {
+                type_constructor_from_modules(
+                    this.compiler.project_compiler.get_importable_modules(),
+                    t,
+                )
+            })) {
```
Could we instead pass in the importable modules? That would make this API simpler, it would remove some of the extra work done in the cases when we don't use the type constructor information to find the location, and it would avoid having to pattern match on the node twice (once in definition_location, once in annotation_type).

lpil commented on compiler-core/src/metadata/tests.rs:
```diff
@@ -324,6 +329,7 @@ fn module_with_type_links() {
                     module: "a".into(),
                     parameters: vec![],
                     deprecation: Deprecation::NotDeprecated,
+                    documentation: Some("documentation".into()),
```
Could we have a dedicated test for the documentation and also have some tests for the origin too please. Thanks

lpil commented on compiler-core/src/type_.rs:
```diff
@@ -134,6 +134,33 @@ impl Type {
+            Self::Var { type_, .. } => type_.borrow().inner_type(),
+            _ => Some(Arc::new(self.clone())),
+        }
+    }
```
I think this is called collapse_type_links elsewhere in the compiler and I think it doesn't clone the type if it is not a link.

Acepie replied:
I spent more time than i want to admit trying to figure out how to do the recursion on the inner `Arc<RefCell<TypeVar>>` and it looks like switching to `collapse_links` actually solves that much nicer than the current implementation. Thanks for pointing this out

lpil commented on compiler-core/src/language_server/engine.rs:
```diff
@@ -225,6 +230,8 @@ where
 
                 Located::Arg(_) => None,
+
+                Located::Annotation(_, _) => Some(this.completion_types(module)),
```
Let's have some tests for this new functionality or remove it.

Acepie replied:
Yup sounds good :D for some reason I thought this was covered by existing tests

lpil reviewed (commented):
Fab! Really like this, will be a huge improvement to the language server.

There's a few tests I would like to have added here (autocompletions for the various places that annotations can happen) and I'd like to go back to how the definition location method worked before. See the note inline.

Thank you so much!!

lpil reviewed (approved):
Beautiful! Thank you so much!!!
